README.md
# MediaServer for mongodb GridFS Documentation

This server intended to serve files from mongodb GridFS.

## How it works

This server works as HTTP server and expose two endpoints:

* `/ping` respond `OK` with status code 200
* `/:filename`
  * if file with `:filename` finded in connected GridFS:
      * send finded file over HTTP and set status code 200
      * in case of malformed `Range` header it return error and appropriate status code
      * in case of error during reading file from connected GridFS it return error and appropriate status code
*  if file with name `:filename` not exists in connected GridFS
      * return status code 404 Not Found

It also support `Range` header and serve requested chunks of file.

All events such as: start server, stopping server, connection to GridFS, http requests, http responses, errors
are logged in syslog with its own severity.

It accepts startup commandline options which are explained in `Running` section.

## Development and deployment

You need installed and fully worked Go compiler.

In order to deploy you need have ruby installed and mina gem

    gem i mina

To compile mserv from source you need:

    make

After sucesfull compilation directory build/ will contain linux and osx executables

To check what you can do with this script you need to run:

    mina tasks

And read output

To deploy server you need do:

    make
    mina server:deploy

## System requirements

Local development: Go compiler with installed cross-compile packages.

Target server: linux 64bit

It need to have /etc/mserv directory with config file:

    mserv.config.yaml

It is a configuration file which mserv use to set port, mongodb connections and so on

    port: 9876
    run_us:
    cpu_use: 4
    mongodb:
        user: etv_import
        password: PaSSword
        hosts:
            - 54.235.213.159:27017
            - 54.235.213.160:27017
        database: classic
        fs: media

All options are self-explained by their names

## Deploy

## Using ruby library `mina`

## Using shell

    # copy binary
    rsync -avz build/mserv-v1.3-64-linux pubapi@54.235.213.159/home/pubapi/mediaserver/

    #!/usr/bin/env bash
    # Executing the following via 'ssh pubapi@54.235.213.159 -t':
    #
    pkill -SIGINT -f mserv 2>/dev/null
    rm -f /home/pubapi/mediaserver/mserv*-linux
    nohup /home/pubapi/mediaserver/mserv-v1.3-64-linux /etc/mserv/mserv.config.yaml 2>/dev/null &

## Running

    > ./mserv # with command line configuration parameters, see below

    > ./mserv path/to/mserv.config.yaml


#### Example of how to run media server

    > ./mserv mserv.config.yaml
       serving on localhost:8080
       utilizing 8 CPU
       Media server started!


CHANGES.md
# 10 Sep 2016

* Update mgo to latest version, in order to proper work with mongo ver > 3.x

# 07 Sep 2014

* Fix issue with serving some of files adding Seek for the end of file before streaming

# 11 Feb 2014

* Add Cache-Control Header into response
* Add ETag Header into response
* Upgrade Go version and compilation

# 28 june 2013

* Add multiplatform compilation
* Add mina ruby gem as generator of deploy bash script
* Written deploy script

# 24 June 2013

* Command-line parsing changed in favor of yaml configuration file
* Added sample yaml config file


config/deploy.rb
set :user, 'pubapi'    # Username in the server to SSH to.
set :project_name, 'mediaserver'
set :domain,      '54.235.213.159'
set :deploy_to,   "/home/#{user}/#{project_name}"


# Put any custom mkdir's in here for when `mina setup` is ran.
# For Rails apps, we'll make some of the shared paths that are shared between
# all releases.
task :setup => :environment do
  queue! %[mkdir -p "#{deploy_to}"]
  queue! %[chmod g+rx,u+rwx "#{deploy_to}"]

  queue! %[mkdir -p "#{deploy_to}/shared"]
  queue! %[chmod g+rx,u+rwx "#{deploy_to}/shared"]
end

desc "Deploys the current version to the server."
task :deploy => :environment do
  deploy do
    to :launch do
      invoke :'server:restart'
    end
  end
end

def mserv_name
  (`cd build/ && find *-linux` || raise).strip
end
set :service_name, mserv_name


namespace :exchange do
  desc "Delete remote mserv"
  task :delete_mserv do
    queue! echo_cmd("rm -f #{deploy_to}/mserv*-linux")
  end

  desc "Copy local build to remote server"
  task :copy_mserv do
    puts "copying #{service_name} to remote server..."
    `rsync -avz build/#{service_name} #{user}@#{domain}:#{deploy_to}`
  end

  desc "Replace old mserv with new from local build"
  task :replace_mserv do
    invoke :'server:stop'
    invoke :'exchange:delete_mserv'
    invoke :'exchange:copy_mserv'
  end
end

namespace :server do
  desc "Deploy new mediaserver and restart"
  task :deploy do
    invoke :'exchange:replace_mserv'
    invoke :'server:restart'
  end


  desc "Stop mediaserver"
  task :stop do
    queue! echo_cmd('pkill -SIGINT -f mserv 2>/dev/null')
  end

  desc "Start mediaserver"
  task :start do
    queue! echo_cmd("nohup #{deploy_to}/#{service_name} /etc/mserv/mserv.config.yaml 2>/dev/null &")
  end

  desc "Restart mediaserver"
  task :restart do
    invoke :'server:stop'
    invoke :'server:start'
  end
end


mserv.go
/*
curl -i -H "Range: bytes=0-3" http://localhost:9876/1c75e2bd-156c-3491-9bf0-409a4c238a4a.jpg
*/

package main

import (
    "errors"
    "flag"
    "fmt"
    "github.com/kylelemons/go-gypsy/yaml"
    //"labix.org/v2/mgo"
    "gopkg.in/mgo.v2"
    "log"
    "log/syslog"
    "net/http"
    "os"
    "os/signal"
    "os/user"
    "runtime"
    "time"
    "strconv"
    "strings"
    "syscall"
)

const Author = "Yury Batenko"
const Version = "1.9 / 2016-10-10"

var (
    srv_bin       string
    session       *mgo.Session
    db            *mgo.Database
    db_connect    string
    slog          *syslog.Writer
    http_port     int
    server_user   string
    cpu_cnt       int
    mongo_hosts   string
    mongo_user    string
    mongo_pass    string
    mongo_db      string
    mongo_fs_name string
    err           error
    ping_resp     string
    req_total     uint64
    req_failed    uint64
    boot_time     string
    hostname      string
)

type gridFSHandler struct {
    GFS      *mgo.GridFS
    PathFile string
}

func DEBUG(x interface{}) {
    fmt.Printf("--> type: %T\n", x)
    fmt.Printf("--- value: %v\n", x)
}

func usage() {
    fmt.Fprintf(os.Stderr, "Usage: %s path/to/mserv.config.yaml\n", os.Args[0])
    flag.PrintDefaults()
    os.Exit(2)
}

func getConfig(filename string) {
    config, err := yaml.ReadFile(filename)
    if err != nil {
        log.Fatalf("readfile(%q): %s", filename, err)
    }

    http_port64, err := config.GetInt("port")
    if err != nil {
        http_port = 9876
    } else {
        http_port = int(http_port64)
    }

    server_user, err = config.Get("run_us")
    if err != nil {
        server_user = ""
    }

    cpu_cnt64, err := config.GetInt("cpu_use")
    if err != nil {
        cpu_cnt = runtime.NumCPU()
    } else {
        cpu_cnt = int(cpu_cnt64)
    }

    mh, err := yaml.Child(config.Root, "mongodb.hosts")
    if err != nil {
        fatal(errors.New("MongoDB hosts not defined!"))
    }
    list, _ := mh.(yaml.List)
    conn := []string{}
    for i := 0; i < list.Len(); i++ {
        conn = append(conn, strings.TrimSpace(yaml.Render(list.Item(i))))
    }
    mongo_hosts = strings.Join(conn, ",")

    mongo_user, err = config.Get("mongodb.user")
    mongo_pass, err = config.Get("mongodb.password")

    mongo_db, err = config.Get("mongodb.database")
    if err != nil {
        fatal(errors.New("MongoDB database not defined!"))
    }

    mongo_fs_name, err = config.Get("mongodb.fs")
    if err != nil {
        mongo_fs_name = "fs"
    }
}

func main() {

    empty := func(s string) bool { return len(s) == 0 }

    flag.Usage = usage
    flag.Parse()

    srv_bin = os.Args[0]

    args := flag.Args()
    if len(args) < 1 {
        fmt.Println("Config file is missing")
        version()
    }

    getConfig(args[0])

    if !empty(server_user) {
        setuid(server_user)
    }

    runtime.GOMAXPROCS(cpu_cnt)

    req_total = 0
    req_failed = 0

    slog, err = syslog.New(syslog.LOG_INFO|syslog.LOG_ERR|syslog.LOG_LOCAL0, "[mserv]")
    if err != nil {
        fatal(err)
    }

    srv_addr := fmt.Sprintf("localhost:%d", http_port)
    fmt.Printf("serving on %s\n", srv_addr)
    fmt.Printf("utilizing %d CPU\n", cpu_cnt)

    slog.Info(fmt.Sprintf("Mserv started and serving on: %s utilizing: %d CPU", srv_addr, cpu_cnt))

    // if !empty(mongo_user) && !empty(mongo_pass) {
    //  db_connect = fmt.Sprintf("%s:%s@%s/%s", mongo_user, mongo_pass, mongo_hosts, mongo_db)
    // } else {
        db_connect = fmt.Sprintf("%s/%s", mongo_hosts, mongo_db)
    // }

    fmt.Println(db_connect)

    ConnectToMongo(db_connect)
    defer session.Close()

    http.Handle("/", GridFSServer(db.GridFS(mongo_fs_name), ""))
    http.HandleFunc("/ping", PingHandler)

    fmt.Println("Media server started!\n")

    boot_time = fmt.Sprintf(time.Now().Format(time.RFC3339))
    hostname, _  = os.Hostname()

    setSignalCatchers()

    s := &http.Server{
        Addr:         srv_addr,
        ReadTimeout:  100000,
        WriteTimeout: 0}
    log.Fatal(s.ListenAndServe())

}

//-------- Utilites ------------------

func setSignalCatchers() {
    go func() {
        sigchan := make(chan os.Signal, 2)
        signal.Notify(sigchan, os.Interrupt)
        <-sigchan
        slog.Info(fmt.Sprintf("exit by SIGINT"))
        os.Exit(0)
    }()
    go func() {
        sigchane := make(chan os.Signal, 2)
        signal.Notify(sigchane, os.Kill, syscall.SIGKILL, syscall.SIGQUIT) // Not catched on MacOSX !
        <-sigchane
        slog.Err(fmt.Sprintf("exit by SIGKILL"))
        os.Exit(-1)
    }()
}

func fatal(err error) {
    fmt.Printf("[!] %s: %s\n", srv_bin, err.Error())
    if slog != nil {
        slog.Err(fmt.Sprintf("Fatal error: %s: %s", srv_bin, err.Error()))
    }
    os.Exit(1)
}

func checkFatal(err error) {
    if err != nil {
        fatal(err)
    }
}

func setuid(username string) {
    usr, err := user.Lookup(username)
    checkFatal(err)
    uid, err := strconv.Atoi(usr.Uid)
    checkFatal(err)
    err = syscall.Setreuid(uid, uid)
    checkFatal(err)
    slog.Info(fmt.Sprintf("set owner of process: %s", username))
}

func version() {
    fmt.Printf("%s Go-GridFS server; author: %s version: %s\n", srv_bin, Author, Version)
    os.Exit(0)
}

//========================================================================

func (g *gridFSHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
    req_total += 1
    filename := r.URL.Path[1:]
    slog.Info(fmt.Sprintf("%s %s %s", r.RemoteAddr, r.Method, r.URL))

    file, err := g.GFS.Open(filename)
    if err != nil {
        req_failed += 1
        http.NotFound(w, r)
        slog.Warning(fmt.Sprintf("requested: %s response: 404 Not Exists", filename))
        return
    }

    defer file.Close()

    etag := file.MD5()
    if inm := r.Header.Get("If-None-Match"); inm != "" {
        if inm == etag || inm == "*" {
                h := w.Header()
                delete(h, "Content-Type")
                delete(h, "Content-Length")
                w.WriteHeader(http.StatusNotModified)
                return
            }
    }

    w.Header().Set("Cache-Control", "max-age=86400, public, must-revalidate, proxy-revalidate")
    w.Header().Set("ETag", etag)
    w.Header().Set("Accept-Ranges", "bytes")
    w.Header().Set("Content-Type", file.ContentType())

    // For some reason SOMETIMES we need to seek to the end of file
    // otherwise we get: seeker can't seek error
    _, err = file.Seek(0, os.SEEK_END)

    slog.Info(fmt.Sprintf("Starting upload of %s", filename))
    http.ServeContent(w, r, file.Name(), file.UploadDate(), file)
    slog.Info(fmt.Sprintf("Finished upload of %s", filename))
}

func PingHandler(w http.ResponseWriter, r *http.Request) {
    slog.Info(fmt.Sprintf("ping request"))
    stat := mgo.GetStats()
    e := session.Ping()
    connected := true
    if e != nil { connected = false }
    ping_resp =
`[status]
status=OK

[server]
version=%s
boot_timestamp=%s
hostname=%s

[mongodb]
connected=%t
MasterConns=%d
SlaveConns=%d
SentOps=%d
ReceivedOps=%d
ReceivedDocs=%d
SocketsAlive=%d
SocketsInUse=%d
SocketRefs=%d

[requests]
total=%d
failed=%d`
    str := fmt.Sprintf(ping_resp, Version, boot_time, hostname, connected, stat.MasterConns, stat.SlaveConns, stat.SentOps, stat.ReceivedOps, stat.ReceivedDocs, stat.SocketsAlive, stat.SocketsInUse, stat.SocketRefs, req_total, req_failed)

    w.Header().Set("Content-Type", "text/plain")
    fmt.Fprint(w, str)
}

// Handle server requests, find file and response.
func GridFSServer(gfs *mgo.GridFS, pathFile string) http.Handler {
    return &gridFSHandler{gfs, pathFile}
}

func ConnectToMongo(connection string) {
    var err error
    empty := func(s string) bool { return len(s) == 0 }

    session, err = mgo.Dial(connection)
    if err != nil {
        log.Fatalf("Error connecting to MongoDB '%s' %v", connection, err.Error())
    }

    if !empty(mongo_user) && !empty(mongo_pass) {
        cred := mgo.Credential{
            Username: mongo_user,
            Password: mongo_pass,
            Source: "admin",
            Mechanism: "SCRAM-SHA-1",
        }
        err = session.Login(&cred)
        if err != nil {
            log.Fatalf("Error authenticate to MongoDB '%s' %v", connection, err.Error())
        }
    }

    // session.SetMode(mgo.Monotonic, true)
    // session.SetMode(mgo.Strong, true)
    session.SetMode(mgo.Eventual, true)
    db = session.DB("") // db will be taken from connection url

    mgo.SetStats(true)
    slog.Info(fmt.Sprintf("Connected to mongodb"))
}
